src/checkpoints.rs

```diff
@@ -0,0 +1,302 @@
+        for i in 1..11 {
+            checkpoints.store(i as u64, i, (i - 1) as u64);
+        }
+        // now checkpoints:
+        // 10->9->8->7->6->5->4->3->2->1
+
+        let points = checkpoints.collect(std::usize::MAX, 10);
+        assert_eq!(points.len(), 10);
+        for (i, (parent, data)) in points.iter().enumerate() {
+            assert_eq!(**data, 10 - i);
+            assert_eq!(*parent, 10 - i as u64);
+        }
+
+        let points = checkpoints.collect(0, 10);
+        assert_eq!(points.len(), 0);
+
+        let points = checkpoints.collect(1, 10);
+        assert_eq!(points.len(), 1);
+
+        let points = checkpoints.collect(std::usize::MAX, 11);
+        assert_eq!(points.len(), 0);
+
+        // only one chain, so tip of trunk is latest
+        assert_eq!(checkpoints.trunk(11), Some(10));
+
+        // only one chain, tip trunk is latest
+        assert_eq!(checkpoints.trunk(1), Some(10));
+
+        // zero length ask
+        assert_eq!(checkpoints.trunk(0), None);
+
+        for i in 11..15 {
+            checkpoints.store(i as u64, i, 1 as u64);
+        }
+        // now checkpoints:
+        // 10->9->8->7->6->5->4->3->2->1
+        // 11->1
+        // 12->1
+        // 13->1
+        // 14->1
+
+        assert_eq!(checkpoints.trunk(1), None);
+        assert_eq!(checkpoints.trunk(0), None);
+        assert_eq!(checkpoints.trunk(9), None);
+        assert_eq!(checkpoints.trunk(10), Some(1));
+        assert_eq!(checkpoints.trunk(11), Some(1));
+
+        // test higher trunk
+        let mut checkpoints: Checkpoints<usize> = Default::default();
+
+        // 3->2->1
+        // 4->2->1
+        checkpoints.store(1, 0, 0);
+        checkpoints.store(2, 0, 1);
+        checkpoints.store(3, 0, 2);
+        checkpoints.store(4, 0, 2);
+        assert_eq!(
+            checkpoints.collect(std::usize::MAX, 3),
+            vec![(3u64, &0), (2u64, &0), (1u64, &0)]
+        );
+        assert_eq!(
+            checkpoints.collect(std::usize::MAX, 4),
+            vec![(4u64, &0), (2u64, &0), (1u64, &0)]
+        );
+        assert_eq!(checkpoints.trunk(std::usize::MAX), Some(2));
+    }
+
+    #[test]
+    fn test_checkpoints_invert() {
+        // test higher trunk
+        let mut checkpoints: Checkpoints<usize> = Default::default();
+
+        // 3->2->1
+        // 4->2->1
+        checkpoints.store(1, 0, 0);
+        checkpoints.store(2, 0, 1);
+        checkpoints.store(3, 0, 2);
+        checkpoints.store(4, 0, 2);
+
+        let _trunks: Vec<_> = checkpoints.invert().0.into_iter().collect();
+        // TODO: returns [0, 2] ??
+
+        // 30->20->10
+        // 40->20->10
+        checkpoints.store(10, 0, 0);
+        checkpoints.store(20, 0, 10);
+        checkpoints.store(30, 0, 20);
+        checkpoints.store(40, 0, 20);
+
+        let _trunks: Vec<_> = checkpoints.invert().0.into_iter().collect();
+        // TODO: returns [0, 2, 10, 20, 1] ??
```

Comment:
@aeyakovenko is this the expected behavior?  I don't understand how 1 is now part of trunks